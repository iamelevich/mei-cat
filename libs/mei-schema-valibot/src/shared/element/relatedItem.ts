import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import { AttrFacsimileSchema } from "../../facsimile/attr/facsimile";
import { AttrBiblSchema } from "../attr/bibl";
import { AttrCommonSchema } from "../attr/common";
import { AttrDataPointingSchema } from "../attr/dataPointing";
import { AttrPointingSchema } from "../attr/pointing";
import { AttrTargetEvalSchema } from "../attr/targetEval";
import { ModelBiblLikeSchema } from "../model/biblLike";

/**
 * Base schema with attribute, to simplify types for RelatedItemSchema
 */
const RelatedItemBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrDataPointingSchema.entries,
	...AttrFacsimileSchema.entries,
	...AttrPointingSchema.entries,
	...AttrTargetEvalSchema.entries,

	// Direct attributes
	/**
	 * Describes the relationship between the entity identified by the <gi scheme="MEI" >relatedItem</gi> element and the resource described in the parent element, <abbr>i.e.</abbr>, <gi scheme="MEI">bibl</gi>, <gi scheme="MEI">source</gi> or <gi scheme="MEI" >relatedItem</gi>.
	 */
	"@rel": v.optional(v.string()),
});

/**
 * Contains or references another bibliographic item which is related to the present one.
 * @see https://music-encoding.org/guidelines/v5/elements/relatedItem.html
 */
export const RelatedItemSchema = v.lazy(() =>
	v.intersect([RelatedItemBaseSchema, ModelBiblLikeSchema]),
);

export type RelatedItemData = v.InferOutput<typeof RelatedItemSchema>;
