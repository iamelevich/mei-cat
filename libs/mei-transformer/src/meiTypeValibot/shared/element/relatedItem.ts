import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import { AttrFacsimileSchema } from "../../facsimile";
import {
	AttrBiblSchema,
	AttrCommonSchema,
	AttrDataPointingSchema,
	AttrPointingSchema,
	AttrTargetEvalSchema,
	ModelBiblLikeSchema,
} from "..";

/**
 * Base schema with attribute, to simplify types for RelatedItemSchema
 */
const RelatedItemBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrDataPointingSchema.entries,
	...AttrFacsimileSchema.entries,
	...AttrPointingSchema.entries,
	...AttrTargetEvalSchema.entries,

	// Direct attributes
	/**
	 * Describes the relationship between the entity identified by the <gi scheme="MEI" >relatedItem</gi> element and the resource described in the parent element, <abbr>i.e.</abbr>, <gi scheme="MEI">bibl</gi>, <gi scheme="MEI">source</gi> or <gi scheme="MEI" >relatedItem</gi>.
	 */
	"@rel": v.optional(v.string()),
});

/**
 * Contains or references another bibliographic item which is related to the present one.
 * @see https://music-encoding.org/guidelines/v5/elements/relatedItem.html
 */
export const RelatedItemSchema = v.intersect([
	RelatedItemBaseSchema,
	ModelBiblLikeSchema,
]);

export type RelatedItemData = v.InferOutput<typeof RelatedItemSchema>;
