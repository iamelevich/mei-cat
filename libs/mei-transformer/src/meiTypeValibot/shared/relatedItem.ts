import * as v from "valibot";
import { AnythingObjectSchema } from "../common";

/**
 * Contains a related item.
 * @see https://music-encoding.org/guidelines/v5/elements/relatedItem.html
 */
export const RelatedItemSchema = AnythingObjectSchema; // TODO: implement element

export type RelatedItemData = v.InferOutput<typeof RelatedItemSchema>;
