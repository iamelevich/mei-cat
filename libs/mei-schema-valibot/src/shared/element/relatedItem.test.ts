import { describe, expect, it } from "bun:test";
import * as v from "valibot";
import { RelatedItemSchema } from "./relatedItem";

describe("RelatedItemSchema", () => {
	it("should be defined", () => {
		expect(RelatedItemSchema).toBeDefined();
	});

	it("should parse a relatedItem with bibl containing title and biblScope", () => {
		const parseResult = v.parse(RelatedItemSchema, {
			bibl: {
				title: {
					"#text": "CNB",
					"@xml:id": "title_N2107D",
				},
				biblScope: {
					"#text": "II/325",
					"@xml:id": "biblScope_N2107F",
				},
				"@label": "",
				"@xml:id": "bibl_d1e55065405",
			},
			"@rel": "host",
			"@xml:id": "relatedItem_N21078",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a relatedItem with different relation type", () => {
		const parseResult = v.parse(RelatedItemSchema, {
			bibl: {
				title: {
					"#text": "M & M",
					"@xml:id": "title_N21086",
				},
				biblScope: {
					"#text": "58",
					"@xml:id": "biblScope_N21088",
				},
				"@xml:id": "bibl_d1e55185417",
			},
			"@rel": "isReferencedBy",
			"@xml:id": "relatedItem_N21081",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a relatedItem with minimal bibl", () => {
		const parseResult = v.parse(RelatedItemSchema, {
			bibl: {
				title: {
					"#text": "CNL",
					"@xml:id": "title_93788bc5b3cd",
				},
				"@xml:id": "bibl_9378db1bae0b",
			},
			"@rel": "host",
			"@xml:id": "relatedItem_9378970ef092",
		});

		expect(parseResult).toMatchSnapshot();
	});

	it("should parse a relatedItem with just title", () => {
		const parseResult = v.parse(RelatedItemSchema, {
			bibl: {
				title: {
					"#text": "Reference Work",
					"@xml:id": "title_ref001",
				},
				"@xml:id": "bibl_ref001",
			},
			"@xml:id": "relatedItem_ref001",
		});

		expect(parseResult).toMatchSnapshot();
	});
});
