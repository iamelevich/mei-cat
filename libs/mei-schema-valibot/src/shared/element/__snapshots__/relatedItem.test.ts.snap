// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`RelatedItemSchema should parse a relatedItem with bibl containing title and biblScope 1`] = `
{
  "@rel": "host",
  "@xml:id": "relatedItem_N21078",
  "bibl": {
    "@label": "",
    "@xml:id": "bibl_d1e55065405",
    "biblScope": {
      "#text": "II/325",
      "@xml:id": "biblScope_N2107F",
    },
    "title": {
      "#text": "CNB",
      "@xml:id": "title_N2107D",
    },
  },
}
`;

exports[`RelatedItemSchema should parse a relatedItem with different relation type 1`] = `
{
  "@rel": "isReferencedBy",
  "@xml:id": "relatedItem_N21081",
  "bibl": {
    "@xml:id": "bibl_d1e55185417",
    "biblScope": {
      "#text": "58",
      "@xml:id": "biblScope_N21088",
    },
    "title": {
      "#text": "M & M",
      "@xml:id": "title_N21086",
    },
  },
}
`;

exports[`RelatedItemSchema should parse a relatedItem with minimal bibl 1`] = `
{
  "@rel": "host",
  "@xml:id": "relatedItem_9378970ef092",
  "bibl": {
    "@xml:id": "bibl_9378db1bae0b",
    "title": {
      "#text": "CNL",
      "@xml:id": "title_93788bc5b3cd",
    },
  },
}
`;

exports[`RelatedItemSchema should parse a relatedItem with just title 1`] = `
{
  "@xml:id": "relatedItem_ref001",
  "bibl": {
    "@xml:id": "bibl_ref001",
    "title": {
      "#text": "Reference Work",
      "@xml:id": "title_ref001",
    },
  },
}
`;
